Sources/GRPC/FakeResponseStream.swift

```diff
@@ -0,0 +1,321 @@
+         (.responsePart(.message), .closed):
+      // We can't send messages once we're closing or closed.
+      return .exclude
+
+    case (.responsePart(.trailingMetadata), .idle),
+         (.responsePart(.trailingMetadata), .sending):
+      self.sendState = .closing
+      return .include
+
+    case (.responsePart(.trailingMetadata), .closing),
+         (.responsePart(.trailingMetadata), .closed):
+      // We're already closing or closed.
+      return .exclude
+
+    case (.responsePart(.status), .idle),
+         (.error, .idle),
+         (.responsePart(.status), .sending),
+         (.error, .sending),
+         (.responsePart(.status), .closed),
+         (.error, .closed):
+      // We can only error/close if we're closing (i.e. have already sent trailers which we enforce
+      // from the API in the subclasses).
+      return .exclude
+
+    case (.responsePart(.status), .closing),
+         (.error, .closing):
+      self.sendState = .closed
+      return .include
+    }
+  }
+
+  private enum FilterResult {
+    /// Include the part.
+    case include
+
+    /// Include the part, but prepend the given part.
+    case includeAndPrepend(_ part: StreamEvent)
+
+    /// Exclude the part.
+    case exclude
+  }
+
+  private func writeOrBuffer(_ part: StreamEvent) {
+    switch self.activeState {
+    case .inactive:
+      self.responseBuffer.append(part)
+
+    case .active:
+      self.write(part)
+    }
+  }
+
+  private func write(_ part: StreamEvent) {
+    switch part {
+    case .error(let error):
+      self.channel.pipeline.fireErrorCaught(error)
+
+    case .responsePart(let responsePart):
+      // We tolerate errors here: an error will be thrown if the write results in an error which
+      // isn't caught in the channel. Errors in the channel get funnelled into the transport held
+      // by the actual call object and handled there.
+      _ = try? self.channel.writeInbound(responsePart)
+    }
+  }
+}
+
+// MARK: - Unary Response
+
+/// A fake unary response to be used with a generated test client.
+///
+/// Users typically create fake responses via helper methods on their generated test clients
+/// corresponding to the RPC which they intend to test.
+///
+/// For unary responses users may call one of two functions for each RPC:
+/// - `sendMessage(_:initialMetadata:trailingMetadata:status)`, or
+/// - `sendError(status:trailingMetadata)`
+///
+/// `sendMessage` sends a normal unary response with the provided message and allows the caller to
+/// also specify initial metadata, trailing metadata and the status. Both metadata arguments are
+/// empty by default and the status defaults to one with an 'ok' status code.
+///
+/// `sendError` may be used to terminate an RPC without providing a response. As for `sendMessage`,
+/// the `trailingMetadata` defaults to being empty.
+public class FakeUnaryResponse<Request: GRPCPayload, Response: GRPCPayload>: FakeResponseStream<Request, Response> {
```

Comment:
Making this type public means that this can be used in prod frameworks/apps depending on gRPC as well, not only their test targets. Is that the intended behaviour?
Do correct me if I'm wrong, but I would expect the test-only types to be internal and hence only accessible from (unit) test targets with `@testable import GRPC`.